File: packages/public-api/src/ContributorService.ts
import Fetch from "./Fetch.js";
import { handleResponse } from "./handleResponse.js";
import { Response, Problem, ContributorResponse, AddContributorRequest } from "./Types.js";

export default class ContributorService {
  private readonly path = `contributor`;
  constructor(private readonly fetch: Fetch) {}

  /** Adds a Contributor - can add fields and search results but not make breaking changes. */
  async add(request: AddContributorRequest): Promise<Response<ContributorResponse, Problem>> {
    const response = await this.fetch.post(this.path, request);
    return handleResponse(response, async () => (await response.json()) as ContributorResponse);
  }

  /** Removes a Contributor's access to the List. */
  async remove(account_name: string, list_name: string, contributor_name: string): Promise<Response<void, Problem>> {
    const path = `${this.path}/${encodeURIComponent(account_name)}/${encodeURIComponent(list_name)}/${encodeURIComponent(contributor_name)}`;
    const response = await this.fetch.delete(path);
    return handleResponse(response, () => undefined);
  }
}

File: dotnet/SearchStack.PublicApi/Services/ContributorService.cs
using System.Threading;
using System.Threading.Tasks;
using static SearchStack.PublicApi.Url;

namespace SearchStack.PublicApi;

/// <summary>Contributor add / remove operations.</summary>
public sealed class ContributorService
{
    private const string Path = "contributor";
    private readonly Transport _transport;

    internal ContributorService(Transport transport) => _transport = transport;

    /// <summary>Adds a Contributor - can add fields and search results but not make breaking changes.</summary>
    public Task<Response<ContributorResponse>> AddAsync(AddContributorRequest request, CancellationToken cancellationToken = default) =>
        _transport.PostAsync<ContributorResponse>(Path, request, cancellationToken);

    /// <summary>Removes a Contributor's access to the List.</summary>
    public Task<Response> RemoveAsync(string accountName, string listName, string contributorName, CancellationToken cancellationToken = default) =>
        _transport.DeleteAsync($"{Path}/{Enc(accountName)}/{Enc(listName)}/{Enc(contributorName)}", cancellationToken);
}
